### PR TITLE
Exception in DataProvider doesn't fail test run

### LINKED ISSUE #217: exception in DataProvider doesn't fail test run
If my @DataProvider method throws an exception, the TestNG report correctly shows that exception. However, the exit status from TestNG execution is zero, indicating that all tests passed. If a DataProvider fails, the build should fail as well.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #217

Fixes #217 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/TestNG.java [hunk 1]
```diff
@@ -211,11 +213,22 @@ private void init(boolean useDefaultListeners) {
   }
 
   public int getStatus() {
+    if (m_status == HAS_STARTED) {
+      //we perhaps are running in the JUnit mode. So the ExitCodeListener is not involved.
```
Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: Is it possible to involve ExiteCodeListener in JUnit mode?

File: src/main/java/org/testng/TestNG.java [hunk 2]
```diff
@@ -211,11 +213,22 @@ private void init(boolean useDefaultListeners) {
...
+      //Lets just override the status and return it as 0
+      return SUCCESS;
+    }
     return m_status;
   }
 
   private void setStatus(int status) {
-    m_status |= status;
+    if (m_status == HAS_STARTED || status == SUCCESS) {
```
Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: If some tests are not success but the last one is success, then the testng status will be success, right?
Is it the expected behavior?

File: src/main/java/org/testng/TestNG.java [hunk 3]
```diff
@@ -1999,8 +2012,13 @@ public void onTestFailure(ITestResult result) {
     @Override
     public void onTestSkipped(ITestResult result) {
       setHasRunTests();
-      if ((m_mainRunner.getStatus() & HAS_FAILURE) != 0) {
```
Comment (juherr) [inline src/main/java/org/testng/TestNG.java]:
```java
if ((m_mainRunner.getStatus() & HAS_FAILURE) != 0 || m_mainRunner.getStatus() == SUCCESS) {
    m_mainRunner.setStatus(HAS_SKIPPED);
}
```

looks shorter

File: src/main/java/org/testng/TestNG.java (hunk 2 above)
Comment (juherr) [reply src/main/java/org/testng/TestNG.java]: The test case should be added. We can enforce a successful test with `priority`.

File: src/main/java/org/testng/TestNG.java (hunk 1 above)
Comment (juherr) [reply src/main/java/org/testng/TestNG.java]: At least, can we add a related test?

File: src/main/java/org/testng/TestNG.java [hunk 4]
```diff
@@ -142,14 +142,16 @@
   private final Map<Class<? extends IReporter>, IReporter> m_reporters = Maps.newHashMap();
   private final Map<Class<? extends IDataProviderListener>, IDataProviderListener> m_dataProviderListeners = Maps.newHashMap();
 
+  protected static final int HAS_STARTED = -1;
+  protected static final int SUCCESS = 0;
   protected static final int HAS_FAILURE = 1;
```
Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: Could you try to replace these values by an enum? I think it will highlight some issues in the logic.

**Comment by krmahadevan** [comment]
After a closer analysis, this PR has issues in terms of properly determining the exit status. Closing this PR.

**Pull request closed without merging**